PetrolCar.cpp

```diff
@@ -1,16 +1,36 @@
 #include "PetrolCar.hpp"
 #include <iostream>
- 
+#include "CarInMovementExcept.hpp"
+
 PetrolCar::PetrolCar(PetrolEngine* engine)
-    : engine_(engine)
-{
-    std::cout << __FUNCTION__ << std::endl;
+    : petrolEngine_(engine) {
+}
+
+PetrolCar::~PetrolCar() {
+}
+void PetrolCar::refill() {
+    refuel();
+}
+void PetrolCar::refuel() {
 }
 
-PetrolCar::~PetrolCar()         { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::refuel()        { std::cout << __FUNCTION__ << std::endl; }
+void PetrolCar::changeEngine(Engine* petrolEngine) {
+    if (currentSpeed_ != 0) {
+        throw CarInMovementExcept("Car is moving");
+    } else if (typeid(*petrolEngine) == typeid(PetrolEngine)) {
+        delete petrolEngine_;
+        petrolEngine_ = static_cast<PetrolEngine*>(petrolEngine);
```

Comment:
Similar here. Can't you just take PetrolEngine as parameter?

Reply:
No because change engine is derived from Car and it has to engine type independent